{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " final_len: mean = 2806.4900 std = (20.3050)\n",
      "-------------------\n"
     ]
    }
   ],
   "source": [
    "sars_vars_path = \"/home/salipe/Desktop/genomic-extractor/GREAC/exec2/benchmark_results_hbv.csv\"\n",
    "data = pd.read_csv(sars_vars_path)\n",
    "# for w in [0.0035]:\n",
    "#     filt = data[(data[\"wndwPercent\"] == w)]\n",
    "#     print(f'window:{w}')\n",
    "for metric in [\"final_len\"]:\n",
    "    print(f' {metric}: mean = {data[metric].mean():.4f} std = ({data[metric].std():.4f})')\n",
    "print('-------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gramep = ['AY.3',\n",
    "'AY.25',\n",
    "'AY.44',   \n",
    "'AY.103',  \n",
    "'B.1' ,\n",
    "'B.1.1.7', \n",
    "'B.1.526', \n",
    "'B.1.2', \n",
    "'BA.1',  \n",
    "'BA.1.1',  \n",
    "'BA.1.15',  \n",
    "'BA.1.18' , \n",
    "'BA.2',\n",
    "'BA.2.12.2', \n",
    "'BA.4.6' ,\n",
    "'BA.5.1' ,\n",
    "'BA.5.2',\n",
    "'BA.5.2.1' ,\n",
    "'BA.5.5', \n",
    "'BQ.1.1' ]\n",
    "\n",
    "kevolve=['B.1.1.7',\n",
    "         'B.1.351',\n",
    "         'P.1',\n",
    "         'B.1.617.2',\n",
    "         'B.1.617.1',\n",
    "         'B.1.427','B.1.429',\n",
    "         'B.1.526',\n",
    "         'B.1.525',\n",
    "         'C.37',\n",
    "         'B.1.1.529',\n",
    "         'BA.1.15','BA.1.18','BA.1.1','BA.2.12.1','BA.4.6','BA.5.1','BA.5.2.1','BA.5.2','BA.5.5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "mutations_file = '/home/salipe/Desktop/datasets/mutation_distribution_in_lineages.txt'\n",
    "\n",
    "threshold = 0.99\n",
    "\n",
    "mutations_df = pd.read_table(mutations_file, sep='\\t', usecols=['Lineage',\n",
    "                                                                    'Frequency of mutated sequences per Lineage',\n",
    "                                                                    'Genome position',\n",
    "                                                                    'Ref seq',\n",
    "                                                                    'ut seq'])\n",
    "                                                                    \n",
    "mutations = mutations_df[(mutations_df['Lineage'].isin(sum([gramep],kevolve))) & (mutations_df['Frequency of mutated sequences per Lineage'] >= threshold)] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations.to_csv(\"teste\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "83"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ls = mutations['Genome position'].to_list()\n",
    "ls = list(set(ls))\n",
    "ls.sort()\n",
    "len(ls)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AY.103',\n",
       " 'AY.25',\n",
       " 'AY.3',\n",
       " 'AY.44',\n",
       " 'B.1',\n",
       " 'B.1.1.529',\n",
       " 'B.1.1.7',\n",
       " 'B.1.2',\n",
       " 'B.1.351',\n",
       " 'B.1.427',\n",
       " 'B.1.429',\n",
       " 'B.1.525',\n",
       " 'B.1.526',\n",
       " 'B.1.617.1',\n",
       " 'B.1.617.2',\n",
       " 'BA.1',\n",
       " 'BA.1.1',\n",
       " 'BA.1.15',\n",
       " 'BA.1.18',\n",
       " 'BA.2',\n",
       " 'BA.2.12.1',\n",
       " 'BA.2.12.2',\n",
       " 'BA.4.6',\n",
       " 'BA.5.1',\n",
       " 'BA.5.2',\n",
       " 'BA.5.2.1',\n",
       " 'BA.5.5',\n",
       " 'BQ.1.1',\n",
       " 'C.37',\n",
       " 'P.1'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(sum([gramep],kevolve))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DENV annotation file positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>1</td>\n",
       "      <td>94</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>95</td>\n",
       "      <td>10273</td>\n",
       "      <td>CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>95</td>\n",
       "      <td>436</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>95</td>\n",
       "      <td>394</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>437</td>\n",
       "      <td>934</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>437</td>\n",
       "      <td>709</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>710</td>\n",
       "      <td>934</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>935</td>\n",
       "      <td>2419</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>2420</td>\n",
       "      <td>3475</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>3476</td>\n",
       "      <td>4129</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>4130</td>\n",
       "      <td>4519</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>4520</td>\n",
       "      <td>6376</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>6377</td>\n",
       "      <td>6757</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>6758</td>\n",
       "      <td>6826</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>6827</td>\n",
       "      <td>7573</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>7574</td>\n",
       "      <td>10270</td>\n",
       "      <td>mature_protein_region_of_CDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>10274</td>\n",
       "      <td>10735</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>10320</td>\n",
       "      <td>10735</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>10393</td>\n",
       "      <td>10735</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>10463</td>\n",
       "      <td>10735</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC_001477.1</td>\n",
       "      <td>10551</td>\n",
       "      <td>10735</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seqid  start    end                          type\n",
       "0   NC_001477.1      1     94                five_prime_UTR\n",
       "1   NC_001477.1     95  10273                           CDS\n",
       "2   NC_001477.1     95    436  mature_protein_region_of_CDS\n",
       "3   NC_001477.1     95    394  mature_protein_region_of_CDS\n",
       "4   NC_001477.1    437    934  mature_protein_region_of_CDS\n",
       "5   NC_001477.1    437    709  mature_protein_region_of_CDS\n",
       "6   NC_001477.1    710    934  mature_protein_region_of_CDS\n",
       "7   NC_001477.1    935   2419  mature_protein_region_of_CDS\n",
       "8   NC_001477.1   2420   3475  mature_protein_region_of_CDS\n",
       "9   NC_001477.1   3476   4129  mature_protein_region_of_CDS\n",
       "10  NC_001477.1   4130   4519  mature_protein_region_of_CDS\n",
       "11  NC_001477.1   4520   6376  mature_protein_region_of_CDS\n",
       "12  NC_001477.1   6377   6757  mature_protein_region_of_CDS\n",
       "13  NC_001477.1   6758   6826  mature_protein_region_of_CDS\n",
       "14  NC_001477.1   6827   7573  mature_protein_region_of_CDS\n",
       "15  NC_001477.1   7574  10270  mature_protein_region_of_CDS\n",
       "16  NC_001477.1  10274  10735               three_prime_UTR\n",
       "17  NC_001477.1  10320  10735                          exon\n",
       "18  NC_001477.1  10393  10735                          exon\n",
       "19  NC_001477.1  10463  10735                          exon\n",
       "20  NC_001477.1  10551  10735                          exon"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "denv = '/home/salipe/Desktop/datasets/denv/annotation.gff3'\n",
    "\n",
    "gff = pd.read_csv(denv, sep=\"\\t\", comment=\"#\", header=None,\n",
    "                  names=[\"seqid\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"phase\", \"attributes\"])\n",
    "\n",
    "# Filtra genes (ou qualquer outro tipo)\n",
    "genes = gff[gff[\"type\"].isin([\"exon\",\"CDS\",\"mature_protein_region_of_CDS\",\"five_prime_UTR\",\"three_prime_UTR\"])]\n",
    "\n",
    "# Cria lista de dicionários\n",
    "dados = []\n",
    "tupples = []\n",
    "for idx, row in genes.iterrows():\n",
    "    atributos = row['attributes']\n",
    "    \n",
    "    # Opcional: extrai ID e Name dos atributos\n",
    "    attr_dict = dict(item.split(\"=\") for item in atributos.split(\";\") if \"=\" in item)\n",
    "    \n",
    "    tupples.append((row[\"start\"],row[\"end\"]))\n",
    "\n",
    "    dados.append({\n",
    "        \"seqid\": row[\"seqid\"],\n",
    "        \"start\": row[\"start\"],\n",
    "        \"end\": row[\"end\"],\n",
    "        \"type\": row[\"type\"]\n",
    "    })\n",
    "df = pd.DataFrame(dados)\n",
    "df\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
